Plugins/NativeEngine/Source/NativeEngine.cpp

```diff
@@ -294,6 +294,140 @@ namespace Babylon
             texture->Height = height;
         }
+
+        class IndexBufferData
+        {
+        public:
+            IndexBufferData(Napi::TypedArray bytes, uint16_t flags, bool dynamic)
+                : Dynamic{dynamic}
+            {
+                const bgfx::Memory* memory = bgfx::copy(bytes.As<Napi::Uint8Array>().Data(), bytes.ByteLength());
+
+                if (Dynamic)
+                {
+                    m_handleIndex = bgfx::createDynamicIndexBuffer(memory, flags).idx;
+                }
+                else
+                {
+                    m_handleIndex = bgfx::createIndexBuffer(memory, flags).idx;
+                }
+            }
+
+            ~IndexBufferData()
+            {
+                if (Dynamic)
+                {
+                    bgfx::destroy(bgfx::DynamicIndexBufferHandle{m_handleIndex});
+                }
+                else
+                {
+                    bgfx::destroy(bgfx::IndexBufferHandle{m_handleIndex});
+                }
+            }
+
+            void Update(Napi::TypedArray bytes, uint32_t startingIdx)
+            {
+                const bgfx::Memory* memory = bgfx::copy(bytes.As<Napi::Uint8Array>().Data(), static_cast<uint32_t>(bytes.ByteLength()));
+                bgfx::update(bgfx::DynamicIndexBufferHandle{m_handleIndex}, startingIdx, memory);
+            }
+
+            std::variant<bgfx::IndexBufferHandle, bgfx::DynamicIndexBufferHandle> Handle() const
+            {
+                if (Dynamic)
+                {
+                    return bgfx::DynamicIndexBufferHandle{m_handleIndex};
+                }
+                else
+                {
+                    return bgfx::IndexBufferHandle{m_handleIndex};
+                }
+            }
+
+            const bool Dynamic{};
+
+        private:
+            uint16_t m_handleIndex{bgfx::kInvalidHandle};
+        };
+
+        class VertexBufferData
+        {
+        public:
+            VertexBufferData(const Napi::Uint8Array& bytes, bool dynamic)
+                : Dynamic{dynamic}
+                , m_bytes{bytes.Data(), bytes.Data() + bytes.ByteLength()}
+            {
+            }
+
+            ~VertexBufferData()
+            {
+                if (m_handleIndex != bgfx::kInvalidHandle)
+                {
+                    if (Dynamic)
+                    {
+                        bgfx::destroy(bgfx::DynamicVertexBufferHandle{m_handleIndex});
+                    }
+                    else
+                    {
+                        bgfx::destroy(bgfx::VertexBufferHandle{m_handleIndex});
+                    }
+                }
+            }
+
+            void EnsureFinalized(Napi::Env env, const bgfx::VertexLayout& layout)
+            {
+                if (m_handleIndex != bgfx::kInvalidHandle)
+                {
+                    return;
+                }
+
+                const bgfx::Memory* bytes = bgfx::copy(m_bytes.data(), static_cast<uint32_t>(m_bytes.size()));
```

Comment:
My initial version was that way, but there are times when `Update()` will get called before `EnsureFinalized()`, which is an allowed behavior. When that happens, the intermediate value stored by the `VertexBufferData` needs to be discarded and replaced with a new one because we can't directly call `bgfx::update()` since the handle isn't valid yet. (This is the first conditional case in `Update()`.) `bgfx::Memory` objects don't have a contract for getting rid of them except to pass them into `bgfx` so that they can be used, then freed; so I didn't see a clean way to proactively make one and decide later that I didn't want to actually give it to `bgfx`. This does end up being an extra copy operation, but only in order to avoid using `bgfx::Memory` in a way it doesn't seem to be designed to support.

Reply:
You can use bgfx::makeRef to control the lifetime?